src/main/scala/akka/persistence/jdbc/query/scaladsl/JdbcReadJournal.scala

```diff
@@ -135,39 +136,96 @@ class JdbcReadJournal(config: Config)(implicit val system: ExtendedActorSystem)
   override def currentEventsByTag(tag: String, offset: Offset): Source[EventEnvelope, NotUsed] =
     currentEventsByTag(tag, offset.value)
 
-  private def currentJournalEventsByTag(tag: String, offset: Long, max: Long): Source[(PersistentRepr, Set[String], JournalRow), NotUsed] = {
+  private def currentJournalEventsByTag(tag: String, offset: Long, max: Long, latestOrdering: MaxOrderingId): Source[EventEnvelope, NotUsed] = {
+    if (latestOrdering.maxOrdering < offset) Source.empty
+    else {
+      readJournalDao.eventsByTag(tag, offset, latestOrdering.maxOrdering, max)
+        .mapAsync(1)(Future.fromTry)
+        .mapConcat {
+          case (repr, _, row) =>
+            adaptEvents(repr).map(r => EventEnvelope(Sequence(row.ordering), r.persistenceId, r.sequenceNr, r.payload))
+        }
+    }
+  }
+
+  /**
+   * @param terminateAfterOffset If None, the stream never completes. If a Some, then the stream will complete once a
+   *                             query has been executed which might return an event with this offset (or a higher offset).
+   *                             The stream may include offsets higher than the value in terminateAfterOffset, since the last batch
+   *                             will be returned completely.
+   */
+  private def eventsByTag(tag: String, offset: Long, terminateAfterOffset: Option[Long]): Source[EventEnvelope, NotUsed] = {
+
     import akka.pattern.ask
-    implicit val askTimeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
-    Source.fromFuture(orderingActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId])
-      .flatMapConcat { latestOrdering =>
-        if (latestOrdering.maxOrdering < offset) Source.empty
-        else readJournalDao.eventsByTag(tag, offset, latestOrdering.maxOrdering, max)
-      }
-      .mapAsync(1)(Future.fromTry)
+    implicit val askTimeout: Timeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
+
+    sealed trait FlowControl
+
+    /** Keep querying - used when we are sure that there is more events to fetch */
+    object Continue extends FlowControl
+
+    /**
+     * Keep querying with delay - used when we have consumed all events,
+     * but want to poll for future events
+     */
+    object ContinueDelayed extends FlowControl
+
+    /** Stop querying - used when we reach the desired offset  */
+    object Stop extends FlowControl
+
+    Source.unfoldAsync[(Long, FlowControl), Seq[EventEnvelope]]((offset, Continue)) {
+
+      case (from, control) =>
+
+        val batchSize = readJournalConfig.maxBufferSize
+
+        def retrieveNextBatch() = {
+
+          for {
+            queryUntil <- journalSequenceActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId]
+            xs <- currentJournalEventsByTag(tag, from, batchSize, queryUntil).runWith(Sink.seq)
+          } yield {
+
+            val offsetsOnly = xs.map(_.offset.value)
+
+            val control =
+              terminateAfterOffset
```

Comment:
A simple pattern match on `terminateAfterOffset` might make this easier to read

Reply:
True. I had another version using `collect`, but I wasn't convinced of it. 

I will try a simple pattern match as suggested